internal/database/unit_tests.go

```diff
@@ -0,0 +1,127 @@
+// defines a number of tests for a type implementing the Database interface
+
+// TODO: this whole test suite could be a lot more terse if we had functions
+// that did ExpectFail, Try, ExpectEqual, etc.
+
+package database
+
+import (
+	"math"
+	"testing"
+
+	"github.com/gpuctl/gpuctl/internal/uplink"
+)
+
+type unitTest struct {
+	Name string
+	F    func(t *testing.T, db Database)
+}
+
+// a list of tests that implementations of the Database interface should pass
+var UnitTests = [...]unitTest{
+	{"DatabaseStartsEmpty", databaseStartsEmpty},
+	{"AppendingFailsIfMachineMissing", appendingFailsIfMachineMissing},
+	{"AppendedDataPointsAreSaved", appendedDataPointsAreSaved},
+	{"MultipleHeartbeats", multipleHeartbeats},
+}
+
+// fake data for adding during tests
+var fakeData = uplink.GPUStats{Name: "GT 1030", Brand: "NVidia",
+	DriverVersion: "v1.4.5", MemoryTotal: 4, MemoryUtilisation: 25.4,
+	GPUUtilisation: 63.5, MemoryUsed: 1.24, FanSpeed: 35.2, Temp: 54.3}
+
+// functions for approximately comparing floats and data structs
```

Comment:
Can we directly override float equality? Unsure if this is possible in Go.

Reply:
Looks like I can use testify for that